hpycc/get.py

```diff
@@ -340,47 +373,60 @@ def get_thor_file(connection, thor_file, max_workers=15, chunk_size=10000,
     schema_str = wuresultresponse["Result"]["XmlSchema"]["xml"]
 
     # get the schema as named tuples of (name, is_set, type)
-    schema = parse_schema_from_xml(schema_str)
-
+    schema, cols = parse_schema_from_xml(schema_str, dtype)
     num_rows = wuresultresponse["Total"]
 
-    # if there are no rows to go and get, we should return an empty dataframe
-    if not num_rows:
-        cols = [c.name for c in schema]
+    if chunk_size is None:  # Automagically optimise. TODO: we could use width too.
+        suggested_size = ceil(num_rows/max_workers)
+        chunk_size = num_rows if suggested_size < 10000 else suggested_size  # Don't chunk small stuff.
+        chunk_size = 325000 if suggested_size > 325000 else chunk_size  # More chunks than workers for big stuff.
+
+    if low_mem:  # Make a temp dir and a blank file to write to
+        temp_dir = tempfile.TemporaryDirectory(dir=temp_dir)
+        temp_file = temp_dir.name + '\\hpycc_temp.csv'
+        pd.DataFrame(columns=cols).to_csv(temp_file, mode='w', index=False)
+    else:
+        temp_file = None
+
+    if not num_rows:  # if there are no rows to go and get, we should return an empty dataframe
         return pd.DataFrame(columns=cols)
 
     chunks = filechunker.make_chunks(num_rows, chunk_size)
-
     with ThreadPoolExecutor(max_workers=max_workers) as executor:
         futures = [
-            executor.submit(connection.get_logical_file_chunk, thor_file,
-                            start_row, n_rows, max_attempts, max_sleep)
+            executor.submit(connection.get_logical_file_chunk, thor_file, start_row,
+                            n_rows, max_attempts, max_sleep, min_sleep, temp_file, cols)
             for start_row, n_rows in chunks
         ]
 
-        finished, _ = wait(futures)
-
-    results = [i.result() for i in finished]
-    flat_list = (item for sublist in results for item in sublist)
-
-    df = pd.DataFrame(flat_list)
+        results = [i.result() for i in futures]  # Wait and exception check too
```

Comment:
so here we can use use the fact that get_logical_file_chunk shold return a json, and write to a load of jsons. we can then read in jsons and update the dicts. this sounds memory efficient and nice?

Reply:
How then to make the dicts a df?

Reply:
We could write each json to a file as we get it. We can then read in these jsons and append them to a list. so we effectively have a list of chunks. combine the sublists and just turn to dataframe?

Reply:
you would have the data in RAM twice though. Once for the dict, once for the df. My thoughts are that low mem should be seek to optimise this wherever possible.

Reply:
is a dict going to have that large a RAM impact? or we could read them in one by one, turn to a df and append to a master df csv. Actually, this approach sounds pretty good?

Reply:
In the end we took the output dicts then appended them to an output dict and deleted as they were produced. Theoretically you could ask for a file bigger than ram now and break it but....why would you want to.